openproblems/tasks/spatial_decomposition/datasets/destvi/utils.py

```diff
@@ -0,0 +1,262 @@
+"""
+From:
+https://github.com/romain-lopez/DestVI-reproducibility/blob/master/simulations/
+"""
+from ...utils import merge_sc_and_sp
+from numba import jit
+from openproblems.data.utils import loader
+from pathlib import Path
+from scipy.spatial.distance import pdist
+from scipy.spatial.distance import squareform
+from sklearn.cluster import AgglomerativeClustering
+from sklearn.decomposition import PCA
+from sklearn.neighbors import kneighbors_graph
+from typing import Optional
+
+import anndata
+import numpy as np
+import pandas as pd
+
+
+def categorical(p, n_samples):
+    size = list(p.shape[:-1])
+    size.insert(0, n_samples)
+    return (p.cumsum(-1) >= np.random.uniform(size=size)[..., None]).argmax(-1).T
+
+
+@jit(nopython=True)
+def get_mean_normal(cell_types, gamma, mean_, components_):  # pragma: no cover
+    """Util for preparing the mean of the normal distribution.
+
+    cell_types: (n_spots, n_cells)
+    gamma: (n_spots, n_cells, n_latent)
+
+    return: samples: (n_spots, n_cells, n_genes)
+    """
+    # extract shapes
+    n_spots = gamma.shape[0]
+    n_cells = gamma.shape[1]
+    n_genes = components_[0].shape[1]
+
+    mean_normal = np.zeros((n_spots, n_cells, n_genes))
+    for spot in range(n_spots):
+        for cell in range(n_cells):
+            mean_normal[spot, cell] = mean_[cell_types[spot, cell]]
+            c = components_[cell_types[spot, cell]]
+            g = np.expand_dims(gamma[spot, cell], 0)
+            mean_normal[spot, cell] += np.dot(g, c)[0]
+    return mean_normal
+
+
+@loader(
+    data_url="https://github.com/romain-lopez/DestVI-reproducibility/",
+    data_reference="https://doi.org/10.1038/s41587-022-01272-8",
+)
+def generate_synthetic_dataset(
+    test: bool,
+    lam_ct: float = 0.1,
+    temp_ct: float = 1.0,
+    lam_gam: float = 0.5,
+    sf_gam: float = 15.0,
+    bin_sampling: float = 1.0,
+    ct_study: int = 0,
+    grid_size: Optional[int] = None,  # size of spatial grid
+    K_sampled: Optional[int] = None,  # cells sampled for each spot
+    seed: int = 0,
+):
+    import torch
+
+    np.random.seed(seed)
+    torch.manual_seed(seed)
+
+    # parameters
+    K = 100
+    if test:
```

Comment:
If test, you'll also need to set `n_obs <= 600`

Reply:
done